pkg/controller/vsphere/reconciler_test.go

```diff
@@ -816,6 +826,171 @@ func TestGetNetworkDevicesMultiDatacenter(t *testing.T) {
+		},
+	}
+
+	poolGroup := "ipam.test.io"
+	addressClaim := []machinev1.NetworkDeviceSpec{
+		{
+			AddressesFromPools: []machinev1.AddressesFromPool{
+				{
+					Name:     "test-pool",
+					Group:    poolGroup,
+					Resource: "ippools",
+				},
+			},
+			Nameservers: []string{
+				"192.168.1.100",
+			},
+		},
+	}
+	addressClaimAndIpPool := []machinev1.NetworkDeviceSpec{
+		{
+			AddressesFromPools: []machinev1.AddressesFromPool{
+				{
+					Name:     "test-pool",
+					Group:    poolGroup,
+					Resource: "ippools",
+				},
+			},
+			Gateway:     "192.168.1.1",
+			IPAddrs:     []string{"192.168.1.2/24"},
+			Nameservers: []string{"192.168.1.100"},
+		},
+	}
+
+	ipAddressClaim := ipamalpha1v1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-claim-0-0",
+			Namespace: "openshift-machine-api",
+		},
+		Spec: ipamalpha1v1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				Name:     "test-pool",
+				APIGroup: &poolGroup,
+				Kind:     "ippools",
+			},
+		},
+		Status: ipamalpha1v1.IPAddressClaimStatus{
+			AddressRef: corev1.LocalObjectReference{
+				Name: "test-test-0",
+			},
+		},
+	}
+
+	ipAddress := ipamalpha1v1.IPAddress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-test-0",
+			Namespace: "openshift-machine-api",
+		},
+		Spec: ipamalpha1v1.IPAddressSpec{
+			ClaimRef: corev1.LocalObjectReference{
+				Name: "test-test-0",
+			},
+			PoolRef: corev1.TypedLocalObjectReference{
+				Name: "test-pool",
+			},
+			Address: "192.168.1.11",
+			Prefix:  24,
+			Gateway: "192.168.1.1",
+		},
+	}
+	testCases := []struct {
+		testCase    string
+		networkSpec []machinev1.NetworkDeviceSpec
+		expected    string
+		err         string
+	}{
+		{
+			testCase:    "Valid IPv4 Static IP",
+			networkSpec: ipv4Static,
+			expected:    "ip=192.168.1.2::192.168.1.1:255.255.255.0:::none nameserver=192.168.1.100",
+		},
+		{
+			testCase:    "Valid IPv6 Static IP",
+			networkSpec: ipv6Static,
+			expected:    "ip=[2001::2]::[2001::1]:64:::none nameserver=[2001::100]",
+		},
+		{
+			testCase:    "Valid dual stack",
+			networkSpec: dualStackStatic,
+			expected:    "ip=192.168.1.2::192.168.1.1:255.255.255.0:::none ip=[2001::2]:::64:::none nameserver=192.168.1.100",
+		},
+		{
+			testCase:    "IPAM Allocated address",
+			networkSpec: addressClaim,
+			expected:    "ip=192.168.1.11::192.168.1.1:255.255.255.0:::none nameserver=192.168.1.100",
+		},
+		{
+			testCase:    "IPAM Allocated address and ipAddrs",
+			networkSpec: addressClaimAndIpPool,
+			expected:    "ip=192.168.1.2::192.168.1.1:255.255.255.0:::none ip=192.168.1.11::192.168.1.1:255.255.255.0:::none nameserver=192.168.1.100",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.testCase, func(t *testing.T) {
+			client := fake.NewClientBuilder().WithScheme(scheme.Scheme).WithRuntimeObjects(
+				&ipAddressClaim,
+				&ipAddress,
+			).Build()
+
+			machineScope := machineScope{
+				Context: context.TODO(),
```

Comment:
Use `context.Background()` in tests as we are never going to actually change this